[wip][train] update Ray Train documentation (quickstart, user guides) (#38287)
ray-project/ray: doc/source/train/getting-started-pytorch.rst

.. _train-pytorch:

Getting Started with PyTorch
============================

This tutorial will walk you through the process of converting an existing PyTorch script to use Ray Train.

By the end of this, you will learn how to:

1. Configure your model so that it runs distributed and is placed on the correct CPU/GPU device.

> 2. Configure your dataset so that it is sharded across the workers and place data on the correct CPU/GPU device.

Reviewer: ```suggestion
2. Configure your dataloader so that it is sharded across the workers and place data on the correct CPU/GPU device.
```

            
            # [4] Configure scaling and resource requirements.
            scaling_config = ScalingConfig(num_workers=2, use_gpu=True)

            # [5] Launch distributed training job.
            trainer = TorchTrainer(train_func, scaling_config=scaling_config)
            result = trainer.fit()

Now, let's get started!

Setting up your training function
---------------------------------

First, you'll want to update your training code to support distributed training. 
You can begin by wrapping your code in a function:

.. code-block:: python

    def train_func(config):
        # Your PyTorch training code here.

This function will be executed on each distributed training worker.

Setting up your model
^^^^^^^^^^^^^^^^^^^^^

Use the :func:`ray.train.torch.prepare_model` utility function. This will:

1. Move your model to the right device.
2. Wrap it in ``DistributedDataParallel``.

.. code-block:: diff

    -from torch.nn.parallel import DistributedDataParallel
    +import ray.train.torch

     def train_func(config): 

         ...

         # Create model.
         model = ...

         # Set up distributed training and device placement.

>     -    device_id = ... # Your logic to get the right device.

Reviewer: I love this, the previous `torch.device(cuda:{localrank})` is too complex

Author: Yeah, it was technically the right way to do it (in Ray Train) but honestly so confusing for someone new.

    +import ray.train.torch

     def train_func(config):

         ...

         dataset = ...
         
         data_loader = DataLoader(dataset, batch_size=worker_batch_size)
    -    data_loader = DataLoader(dataset, batch_size=worker_batch_size, sampler=DistributedSampler(dataset)) 
    +    data_loader = ray.train.torch.prepare_data_loader(data_loader)

         for X, y in data_loader:
    -        X = X.to_device(device)
    -        y = y.to_device(device)

         ...

.. tip::
    Keep in mind that ``DataLoader`` takes in a ``batch_size`` which is the batch size for each worker.
    The global batch size can be calculated from the worker batch size (and vice-versa) with the following equation:

    .. code-block:: python

        global_batch_size = worker_batch_size * ray.train.get_context().get_world_size()

Reporting metrics and checkpoints
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To monitor progress, you can report intermediate metrics and checkpoints using the :func:`ray.train.report` utility function.

.. code-block:: diff

    +import ray.train
    +from ray.train import Checkpoint

     def train_func(config):

         ...

> 

Reviewer: ```suggestion
         torch.save(model.state_dict(), f"{checkpoint_dir}/model.pth"))
```

    +    metrics = {"loss": loss.item()} # Training/validation metrics.

>     +    checkpoint = Checkpoint.from_directory(checkpoint_dir) # A directory with checkpoint files.

Reviewer: ```suggestion
    +    checkpoint = Checkpoint.from_directory(checkpoint_dir) # Build a Ray Train checkpoint from a directory
```
